lib/index.d.ts

```diff
@@ -1300,105 +1355,102 @@ declare namespace planck {
-        (def: PrismaticJointDef & {localAnchorA: Vec2, localAnchorB: Vec2, localAxisA: Vec2}, bodyA: Body, bodyB: Body): PrismaticJoint;
-        (def: PrismaticJointDef & {localAxisA: Vec2}, bodyA: Body, bodyB: Body, anchor: Vec2): PrismaticJoint;
-        (def: PrismaticJointDef | null | undefined, bodyA: Body, bodyB: Body, anchor: Vec2, axis: Vec2): PrismaticJoint;
+        new(def: PrismaticJointDef): PrismaticJoint;
+           (def: PrismaticJointDef): PrismaticJoint;
+
+        new(def: PrismaticJointOpt, bodyA: Body, bodyB: Body, anchor: Vec2, axis: Vec2): PrismaticJoint;
+           (def: PrismaticJointOpt, bodyA: Body, bodyB: Body, anchor: Vec2, axis: Vec2): PrismaticJoint;
 
         TYPE: 'prismatic-joint';
     }
     let PulleyJoint: {
-        new(def: PulleyJointDef & {ratio: number}, bodyA: Body, bodyB: Body, groundA: Vec2, groundB: Vec2, anchorA: Vec2, anchorB: Vec2): PulleyJoint;
-        new(def: PulleyJointDef | null | undefined, bodyA: Body, bodyB: Body, groundA: Vec2, groundB: Vec2, anchorA: Vec2, anchorB: Vec2, ratio: number): PulleyJoint;
-        (def: PulleyJointDef & {ratio: number}, bodyA: Body, bodyB: Body, groundA: Vec2, groundB: Vec2, anchorA: Vec2, anchorB: Vec2): PulleyJoint;
-        (def: PulleyJointDef | null | undefined, bodyA: Body, bodyB: Body, groundA: Vec2, groundB: Vec2, anchorA: Vec2, anchorB: Vec2, ratio: number): PulleyJoint;
+        new(def: PulleyJointDef): PulleyJoint;
+           (def: PulleyJointDef): PulleyJoint;
+
+        new(def: PulleyJointOpt, bodyA: Body, bodyB: Body, groundA: Vec2, groundB: Vec2, anchorA: Vec2, anchorB: Vec2, ratio: number): PulleyJoint;
+           (def: PulleyJointOpt, bodyA: Body, bodyB: Body, groundA: Vec2, groundB: Vec2, anchorA: Vec2, anchorB: Vec2, ratio: number): PulleyJoint;
 
         TYPE: 'pulley-joint';
         MIN_PULLEY_LENGTH: number;
     }
     let RevoluteJoint: {
-        new(def: RevoluteJointDef & {localAnchorA: Vec2, localAnchorB: Vec2}, bodyA: Body, bodyB: Body): RevoluteJoint;
-        new(def: RevoluteJointDef | null | undefined, bodyA: Body, bodyB: Body, anchor: Vec2): RevoluteJoint;
-        (def: RevoluteJointDef & {localAnchorA: Vec2, localAnchorB: Vec2}, bodyA: Body, bodyB: Body): RevoluteJoint;
-        (def: RevoluteJointDef | null | undefined, bodyA: Body, bodyB: Body, anchor: Vec2): RevoluteJoint;
+        new(def: RevoluteJointDef): RevoluteJoint;
+           (def: RevoluteJointDef): RevoluteJoint;
+
+        new(def: RevoluteJointOpt, bodyA: Body, bodyB: Body, anchor: Vec2): RevoluteJoint;
+           (def: RevoluteJointOpt, bodyA: Body, bodyB: Body, anchor: Vec2): RevoluteJoint;
 
         TYPE: 'revolute-joint';
     }
     let RopeJoint: {
-        new(def: RopeJointDef & {localAnchorA: Vec2, localAnchorB: Vec2, bodyA: Body, bodyB: Body}): RopeJoint;
-        new(def: RopeJointDef & {localAnchorA: Vec2, localAnchorB: Vec2}, bodyA: Body, bodyB: Body): RopeJoint;
```

Comment:
So the API is simply not intended to be used with these strange mixed arguments? (I'm just curious)

Reply:
Box2D API expects a definition object to create a joint, and some definitions can be initialized with different params ([see here](https://github.com/erincatto/Box2D/blob/master/Box2D/Dynamics/Joints/b2DistanceJoint.h)).

In Planck I have converted all definitions to simple objects (to make it more convenient to use), so there is no method to call to initialize a definition, but instead I have added those params to joints' constructor. So you can either pass everything in a definition object or pass params to constructor (practically in JS you can mix them but it is not supposed to be used like that).